### PR TITLE
Speed up non-matching interpolation data and add extrapolation parameter

### PR DESCRIPTION
**Pull request by jorgensd**
Following the regression documented in #2847 I propose that we make extrapolation distance an input parameter (padding), which is used to pad the bounding boxes used to find processes with potential collisions.

The new code will first do as the original interpolation data did:
- Given a set of interpolation points (locally to each process) create a padded bounding box on the mesh we want to interpolate to.
- Find what processes has the interpolation points, and distribute data so we can use GJK on those processes to find the first colliding cell.
- Send this cell back to the interpolation point owner

Then the strategy is:
- For points that have not gotten an owner assigned (and are within the padded bounding box),  compute the closest entity  and use this as owner.

@massimiliano-leoni  could you test this branch with your example, setting the padding first to 0, then to 1e-10, 1e-6, 1e-2 etc.?

For performance impact, see: #2862

### CODE REVIEW

File: cpp/dolfinx/geometry/utils.h [hunk 1]
```diff
@@ -847,6 +937,7 @@ determine_point_ownership(const mesh::Mesh<T>& mesh, std::span<const T> points)
   owned_recv_ranks.reserve(recv_offsets.back());
   std::vector<T> owned_recv_points;
   std::vector<std::int32_t> owned_recv_cells;
+
```
Comment (jorgensd) [inline cpp/dolfinx/geometry/utils.h:940]:
```suggestion
```

**Comment by jorgensd** [comment]
This PR would also resolve #2786

File: cpp/dolfinx/geometry/utils.h [hunk 2]
```diff
@@ -758,15 +758,32 @@ determine_point_ownership(const mesh::Mesh<T>& mesh, std::span<const T> points)
...
-  const std::vector<T> squared_distances = squared_distance(
-      mesh, tdim, closest_cells,
-      std::span<const T>(received_points.data(), received_points.size()));
+  std::vector<std::int32_t> cell_indicator(received_points.size() / 3);
+  std::vector<std::int32_t> closest_cells(received_points.size() / 3);
+  for (std::size_t p = 0; p < received_points.size(); p += 3)
+  {
+    std::array<T, 3> point;
+    std::copy_n(std::next(received_points.begin(), p), 3, point.begin());
+    const int colliding_cell = geometry::compute_first_colliding_cell(
```
Comment (garth-wells) [inline cpp/dolfinx/geometry/utils.h:781]: This code isn't easy to read. Could it be simplified?

File: cpp/dolfinx/geometry/utils.h [hunk 3]
```diff
@@ -793,26 +810,105 @@ determine_point_ownership(const mesh::Mesh<T>& mesh, std::span<const T> points)
     std::swap(recv_offsets, send_offsets);
   }
 
+  std::vector<std::int32_t> recv_ranks(recv_offsets.back());
+  MPI_Neighbor_alltoallv(
+      cell_indicator.data(), send_sizes.data(), send_offsets.data(),
+      dolfinx::MPI::mpi_type<std::int32_t>(), recv_ranks.data(),
```
Comment (garth-wells) [inline cpp/dolfinx/geometry/utils.h:816]: `MPI_INT32_T`

File: cpp/dolfinx/geometry/utils.h [hunk 4]
```diff
@@ -793,26 +810,105 @@ determine_point_ownership(const mesh::Mesh<T>& mesh, std::span<const T> points)
...
+      recv_sizes.data(), recv_offsets.data(),
+      dolfinx::MPI::mpi_type<std::int32_t>(), reverse_comm);
```
Comment (garth-wells) [inline cpp/dolfinx/geometry/utils.h:818]: `MPI_INT32_T`

File: cpp/dolfinx/geometry/utils.h [hunk 5]
```diff
@@ -793,26 +810,105 @@ determine_point_ownership(const mesh::Mesh<T>& mesh, std::span<const T> points)
...
+
+  std::vector<std::int32_t> point_owners(points.size() / 3, -1);
+  for (std::size_t i = 0; i < unpack_map.size(); i++)
+  {
+    const std::int32_t pos = unpack_map[i];
+    // Only insert new owner if no owner has previously been found
+    if ((recv_ranks[i] >= 0) && (point_owners[pos] == -1))
```
Comment (garth-wells) [inline cpp/dolfinx/geometry/utils.h:825]: No need for the extra parentheses.

File: cpp/dolfinx/geometry/utils.h [hunk 6]
```diff
@@ -793,26 +810,105 @@ determine_point_ownership(const mesh::Mesh<T>& mesh, std::span<const T> points)
...
+      point_owners[pos] = recv_ranks[i];
+  }
+
+  // Create extrapolation marker for those points already sent to other
+  // process
+  std::vector<std::uint8_t> send_extrapolate(recv_offsets.back());
+  for (std::int32_t i = 0; i < recv_offsets.back(); i++)
+  {
+    const std::int32_t pos = unpack_map[i];
+    send_extrapolate[i] = std::uint8_t(point_owners[pos] == -1);
```
Comment (garth-wells) [inline cpp/dolfinx/geometry/utils.h:835]: Is explicit conversion required?

File: cpp/dolfinx/geometry/utils.h [hunk 7]
```diff
@@ -793,26 +810,105 @@ determine_point_ownership(const mesh::Mesh<T>& mesh, std::span<const T> points)
...
   for (std::size_t i = 0; i < out_ranks.size(); i++)
   {
     for (std::int32_t j = recv_offsets[i]; j < recv_offsets[i + 1]; j++)
     {
       const std::int32_t pos = unpack_map[j];
-      // If point has not been found yet distance is negative. If new
-      // received distance smaller than current distance choose owner.
-      if (auto d = closest_distance[pos]; d < 0 or d > recv_distances[j])
+      auto current_dist = recv_distances[j];
+      // Only update if closer than previous guess and was found
```
Comment (garth-wells) [inline cpp/dolfinx/geometry/utils.h:906]: Remove 'Only"

**Review by garth-wells — DISMISSED** [review]
I think use of the work 'extrapolation' can be tightened up, which will help understanding. 'Extrapolation' is used in the context of geometric operations, which doesn't seem right.

**Comment by massimiliano-leoni** [comment]
I just tried the latest revision in this branch and I can confirm that the performance is again satisfying.
On the flip side, I tried setting the padding to `1` or `1e-1` just to see what happens and I noticed an explosion in RAM usage. In fact, I crashed my own machine :_D

**Comment by jorgensd** [comment]
> I just tried the latest revision in this branch and I can confirm that the performance is again satisfying. On the flip side, I tried setting the padding to `1` or `1e-1` just to see what happens and I noticed an explosion in RAM usage. In fact, I crashed my own machine :_D

Well, that is expected, as you end up with massive padding (tons of points to loop through).
I don't have any good way of counteracting this behavior, and have therefore added warnings/notes that this value should be set on a problem basis, and should likely not be very large.

**Comment by massimiliano-leoni** [comment]
Yes, I guess that's reasonable :)

File: cpp/dolfinx/geometry/utils.h (hunk 2 above)
Comment (jorgensd) [reply cpp/dolfinx/geometry/utils.h:792]: I've added comments to try to address this

**Comment by jorgensd** [comment]
> I think use of the work 'extrapolation' can be tightened up, which will help understanding. 'Extrapolation' is used in the context of geometric operations, which doesn't seem right.

I've removed the usage of extrapolate in the documentation. (Except to explain that it could be used for this).

**Pull request merged**